src/main/kotlin/com/exactpro/th2/woodpecker/api/impl/event/EventGenerator.kt

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2023 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.woodpecker.api.impl.event
+
+import com.exactpro.th2.common.grpc.EventBatch
+import com.exactpro.th2.woodpecker.api.event.IEventGenerator
+import com.exactpro.th2.woodpecker.api.impl.GeneratorSettings
+import com.exactpro.th2.woodpecker.api.impl.event.provider.EventProvider
+import com.exactpro.th2.woodpecker.api.impl.event.provider.MixedEventProvider
+import com.exactpro.th2.woodpecker.api.impl.event.provider.MultiRootEventProvider
+import com.exactpro.th2.woodpecker.api.impl.event.provider.SingleRootEventProvider
+import mu.KotlinLogging
+import kotlin.math.pow
+
+class EventGenerator(
+    val settings: EventGeneratorSettings
+) : IEventGenerator<GeneratorSettings> {
+    private val logger = KotlinLogging.logger { }
+    private var context = Context(settings)
+    private val treeSize = calculateTreeSize()
+
+    override fun onStart(settings: GeneratorSettings?) {
+        settings?.let {
+            context = Context(settings.eventGeneratorSettings)
+            logger.info { "Reset generator settings" }
+        }
+    }
+
+    override fun onStop() {
+        logger.info { "EventGenerator::onStop" }
+    }
+
+    override fun onNext(size: Int): EventBatch {
+        if (treeSize > size) {
+            throw RuntimeException("batch size can't be less than initial event tree size")
+        }
```

Comment:
since we need this mostly for testing purposes I thought that we could keep the logic simple and with minimal changes in core. If wee see a general need for this use case in future we could improve this logic after that. While in development we can just use a bit large batch sizes to avoid issue

Reply:
OK